### ISSUE TITLE
strict-boolean-expressions false positive in react

### ISSUE THREAD

**Issue by pgsandstrom** [palantir/tslint#4748]
### Bug Report

- __TSLint version__: 5.16.0
- __TypeScript version__: 3.3.4000
- __Running TSLint via__: (pick one) CLI

#### TypeScript code being linted

```ts
interface State {
 message?:string
}

const initialState = { message: undefined }

class MyComponent extends React.Component<{}, State> {

  state = initialState

  componentDidMount() {
    this.setState({message: 'hello'})
  }

  render() {
    if(this.state.message) {
      return <div>{this.state.message}</div>
    } else {
      return <div>no message</div>
    }
  }
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "strict-boolean-expressions": [true, "allow-null-union", "allow-undefined-union", "allow-string", "allow-enum", "allow-number", "allow-mix", "ignore-rhs"]
  }
}
```

#### Actual behavior
Fails with message `This type is not allowed in the 'if' condition because it is always falsy. Allowed types are boolean, null-union, undefined-union, string, enum, or number.`

#### Expected behavior
The test should pass. Perhaps properties on extended classes should be exempt from analysis.

**Comment by JoshuaKGoldberg** [comment]
Great issue @pgsandstrom! This is actually a quirk on TypeScript's end: by declaring `state = initialState` instead of `state: State = initialState`, `state` is _not_ inferred to be of type `State`.

```diff
+  state: State = initialState
-  state = initialState
```

This is rather confusing and arguably irritating... but there's nothing TSLint can do about it 😦. I'm not really sure why; it might be something to do with local member declarations overriding the declared template type? You'll want to file an issue on TypeScript if you think this behavior is incorrect.
